api/authentication.go

```diff
@@ -8,138 +8,345 @@ import (
+var (
+	authFnSet = abool.New()
+	authFn    Authenticator
+
+	authTokens     = make(map[string]*AuthToken)
+	authTokensLock sync.Mutex
+
+	// ErrAPIAccessDeniedMessage should be returned by Authenticator functions in
+	// order to signify a blocked request, including a error message for the user.
+	ErrAPIAccessDeniedMessage = errors.New("")
+)
 
 // SetAuthenticator sets an authenticator function for the API endpoint. If none is set, all requests will be permitted.
 func SetAuthenticator(fn Authenticator) error {
 	if module.Online() {
 		return ErrAuthenticationImmutable
 	}
 
-	authFnLock.Lock()
-	defer authFnLock.Unlock()
-
-	if authFn != nil {
+	if authFnSet.IsSet() {
 		return ErrAuthenticationAlreadySet
 	}
 
 	authFn = fn
+	authFnSet.Set()
 	return nil
 }
 
 func authMiddleware(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tracer := log.Tracer(r.Context())
+		token := authenticateRequest(w, r, nil)
+		if token != nil {
+			if _, apiRequest := getAPIContext(r); apiRequest != nil {
+				apiRequest.AuthToken = token
+			}
+			next.ServeHTTP(w, r)
+		}
+	})
+}
 
-		// get authenticator
-		authFnLock.Lock()
-		authenticator := authFn
-		authFnLock.Unlock()
+func authenticateRequest(w http.ResponseWriter, r *http.Request, targetHandler http.Handler) *AuthToken {
+	tracer := log.Tracer(r.Context())
 
-		// permit if no authenticator set
-		if authenticator == nil {
-			next.ServeHTTP(w, r)
-			return
+	// Check if authenticator is set.
+	if !authFnSet.IsSet() {
+		// Return highest available permissions.
+		return &AuthToken{
+			Read:  PermitSelf,
+			Write: PermitSelf,
 		}
+	}
 
-		// check existing auth cookie
-		c, err := r.Cookie(cookieName)
-		if err == nil {
-			// get token
-			validTokensLock.Lock()
-			validUntil, valid := validTokens[c.Value]
-			validTokensLock.Unlock()
-
-			// check if token is valid
-			if valid && time.Now().Before(validUntil) {
-				tracer.Tracef("api: auth token %s is valid, refreshing", c.Value)
-				// refresh cookie
-				validTokensLock.Lock()
-				validTokens[c.Value] = time.Now().Add(cookieTTL)
-				validTokensLock.Unlock()
-				// continue
-				next.ServeHTTP(w, r)
-				return
-			}
+	// Check if request is read only.
+	readRequest := isReadMethod(r.Method)
+
+	// Get required permission for target handler.
+	requiredPermission := PermitSelf
+	if authdHandler, ok := targetHandler.(AuthenticatedHandler); ok {
+		if readRequest {
+			requiredPermission = authdHandler.ReadPermission(r)
+		} else {
+			requiredPermission = authdHandler.WritePermission(r)
+		}
+	}
 
-			tracer.Tracef("api: provided auth token %s is invalid", c.Value)
+	// Check if we need to do any authentication at all.
+	switch requiredPermission {
+	case NotFound:
+		// Not found.
+		tracer.Trace("api: authenticated handler reported: not found")
+		http.Error(w, "Not found.", http.StatusNotFound)
+		return nil
```

Comment:
The handler may be multiplexing requests again, and can already say if the given path even exists.

Reply:
But the handler receives the `http.ResponseWriter` anyway so it can already respond with `http.Error(w, "Not found", http.StatusNotFound)` by itself. Not sure we need an extra way for that (IMHO the handler exists so it should handle the 404 for any "sub-handler" on it's own.